bigframes/core/nodes.py

```diff
@@ -1486,3 +1486,56 @@ def remap_refs(
     ) -> BigFrameNode:
         new_ids = tuple(id.remap_column_refs(mappings) for id in self.column_ids)
         return dataclasses.replace(self, column_ids=new_ids)  # type: ignore
+
+
+# Tree operators
+def top_down(
+    root: BigFrameNode,
+    transform: Callable[[BigFrameNode], BigFrameNode],
+    *,
+    memoize=False,
+    validate=False,
+) -> BigFrameNode:
+    """
+    Perform a top-down transformation of the BigFrameNode tree.
+
+    If memoize=True, recursive calls are memoized within the scope of the traversal only.
+    """
+
+    def top_down_internal(root: BigFrameNode) -> BigFrameNode:
+        return transform(root).transform_children(top_down_internal)
+
+    if memoize:
+        # MUST reassign to the same name or caching won't work recursively
+        top_down_internal = functools.cache(top_down_internal)
+
+    result = top_down_internal(root)
+    if validate:
+        result.validate_tree()
+    return result
+
+
+def bottom_up(
+    root: BigFrameNode,
+    transform: Callable[[BigFrameNode], BigFrameNode],
+    *,
+    memoize=False,
+    validate=False,
+) -> BigFrameNode:
+    """
+    Perform a bottom-up transformation of the BigFrameNode tree.
+
+    If memoize=True, recursive calls are memoized within the scope of the traversal only.
+    """
+
+    def bottom_up_internal(root: BigFrameNode) -> BigFrameNode:
+        return transform(root.transform_children(bottom_up_internal))
```

Comment:
Don't see a difference with top_down. Should bottom_up apply to parents?

Reply:
top down applies `transform_children()` after `transform(root)` that is the only difference.

Reply:
```
def bottom_up_internal(root: BigFrameNode) -> BigFrameNode:
        return transform(root.transform_children(bottom_up_internal))
```
Then this is a mistake?

Reply:
nope, not a mistake. bottom_up transforms the children first. and then transforms the root last.

Reply:
Then bottom up doesn't apply to their parents? How is that "bottom up"?

Reply:
to which parents? its a classic post-order depth-first traversal, all nodes under the root are transformed, and then the root is the final node to be transformed

Reply:
OIC. I was thinking it to start from leaf, but it finds the leafs from the root. Cool.